api/classes/publisher.js

```diff
@@ -79,72 +115,236 @@ function deleteFile(file, root) {
+  });
+}
+
+/**
+ * Exports
+ */
+exports.publish = function publish(project) {
+  var user;
+  var files;
+
+  var publishedFiles;
+  var publishedProject;
+
+  var projectFiles = {};
+  var publishedProjectFiles = {};
+  var toBeCreated = [];
+  var toBeDeleted = [];
+  var toBeUpdated = [];
+
+  return Promise.resolve()
+  .then(function() {
+    return getUserForProject(project)
+    .then(function(u) {
+      user = u;
     });
   })
   .then(function() {
-    return buildUrl(project, user);
+    return getPublishedProject(project)
+    .then(function(oldPublishedProject) {
+      if (oldPublishedProject) {
+        return oldPublishedProject.set({
+          title: project.get('title'),
+          tags: project.get('tags'),
+          description: project.get('description')
+        }).save();
+      }
+      return PublishedProject.forge({
+        title: project.get('title'),
+        tags: project.get('tags'),
+        description: project.get('description')
+      }).save();
+    })
+    .then(function(p) {
+      publishedProject = p;
+    });
   })
-  .catch(function(e) {
-    return Promise.reject(e);
-  });
-}
+  .then(function() {
+    return File.query({
+      where: {
+        project_id: project.get('id')
+      }
+    }).fetchAll()
+    .then(function(f) {
+      if (f.length === 0) {
+        throw 'No files to be published!';
+      }
 
-function getUser(project) {
-  return project.user().query({}).fetch();
-}
+      files = f;
+    });
+  })
+  .then(function() {
+    return PublishedFiles.query({
+      where: {
+        published_id: publishedProject.get('id')
+      }
+    }).fetchAll()
+    .then(function(f) {
+      if (f.length === 0) {
+        return;
+      }
+      publishedFiles = f;
+    });
+  })
+  .then(function() {
+    // We want to organize 'file' objects by id, and
+    // 'publishedFile' objects by the 'file' objects they reference
+    files.forEach(function(file) {
+      projectFiles[file.get('id')] = file;
+    });
+    var filePrimaryKeys = Object.keys(projectFiles);
 
-exports.publish = function(project) {
-  return getUser(project)
-    .then(function(user) {
-      return modifyFiles(uploadFile, project, user);
-    })
-    .then(function(url) {
-      return project.set({
-        publish_url: url
+    // Nothing published yet?
+    if (!publishedFiles) {
+      return filePrimaryKeys.forEach(function(id) {
+        return toBeCreated.push(projectFiles[id]);
+      });
+    }
+
+    // We collect references to each model that needs to
+    // have its 'publishedFile' created or destroyed (respectively)
+    publishedFiles.forEach(function(publishedFile) {
+      if (publishedFile.get('file_id') === null) {
+        toBeDeleted.push(publishedFile);
+        return;
+      }
+      publishedProjectFiles[publishedFile.get('file_id')] = publishedFile;
+    });
+    var publishedFileForeignKeys = Object.keys(publishedProjectFiles);
+
+    filePrimaryKeys.forEach(function(primaryKey) {
```

Comment:
Also, I wonder if you can separate them into functions? Maybe that would make it clearer! Otherwise it seems hard to follow!
e.g. `function projectDeleteDiffs()`, `function projectCreateDiffs()`... or something like that